design-system/src/theme-provider.tsx

```diff
@@ -64,12 +65,26 @@ type ThemeProviderProps = {
 
 const ThemeProvider = (props: ThemeProviderProps) => {
   const parentSelectorRef = useRef(props.parentSelector);
+  const themeRef = useRef<string>();
+  const themeOverridesRef = useRef<Record<string, string>>();
+
   useLayoutEffect(() => {
-    applyTheme({
-      newTheme: props.theme,
-      parentSelector: parentSelectorRef.current,
-      themeOverrides: props.themeOverrides,
-    });
+    // We want to make sure we don't really apply the change when the props
+    // provided include a new object with the same theme overrides
+    // (eg: users providing an inline object as prop to the ThemeProvider)
+    if (
+      themeRef.current !== props.theme ||
+      !isEqual(themeOverridesRef.current, props.themeOverrides)
+    ) {
+      themeRef.current = props.theme;
+      themeOverridesRef.current = props.themeOverrides;
+
+      applyTheme({
+        newTheme: props.theme,
+        parentSelector: parentSelectorRef.current,
+        themeOverrides: props.themeOverrides,
+      });
+    }
   }, [props.theme, props.themeOverrides]);
```

Comment:
Thanks for looking into that!

I was wondering if there is also a way of avoiding passing these values as "dependencies" in the first place 🤔

Reply:
Do you mean removing the effect?

I guess the way it is right now helps us not triggering the effect when users don't provide a `themeOverrides` prop or they don't provide an inline object. If they do, then we have the new check in place.

If we remove the effect, the actual result will be the same, but the check would be run on every render.
Also, we need to apply the theme in the first render.

Maybe if you have some idea in mind you can share it and I can explore it.

Reply:
Sure, I was just thinking about computing (merging) the theme properties outside of the effect and then pass the result into the effect. So the effect is only responsible to apply the value to the DOM but the computation can be done outside of it.

What do you think?

Reply:
I'm sorry but I think I'm not following you. If we compute the theme properties (do you mean creating the `var` that will be applied to the parent selector?) outside the effect, we will be computing those props on every render, even when the original props did not change and we will not need to apply the change.

Perhaps we can arrange a quick call to go over this topic.

Reply:
No problem, I'm just "thinking out loud" here. Sorry for not being too clear.

I was referring more to the fact that the effect should only apply the theme but the computed theme can be calculated before (for example using `useMemo` or something). Still, to ensure referencial equality of the `themeOverrides` (being an object) we need to keep the ref around, as you have already implemented.

Anyway, this is perfectly fine (and better than before) so let's go with this 🤗

Reply:
So, if this is fine, could I have an approval, please 🙏 ?